### DISCUSSION TITLE
what is the best way to use my build func to build model?

### DISCUSSION THREAD

**Discussion by Fei-Wang** [open-mmlab/mmengine#900] [Q&A]
This is my code.
```python3
from mmengine.registry import MODELS as MMENGINE_MODELS
MODELS = Registry('model', build_func=build_model, parent=MMENGINE_MODELS, locations=['myproject.models'])
def build_model(cfg, registry, *args, **kwargs):
    pass
```
but, when I use default runner in myproject, here is the error
 https://github.com/open-mmlab/mmengine/blob/5753cd98e0b791897a2d402b7eb196fb04b4a4eb/mmengine/runner/runner.py#L800
this Models registry is mmengine MODELS, not myproject MODELS, so it can't find my build_func, so what is the best way to use my build func?

**Comment by C1rN09** [comment]
Currently you can't use your own `build_func`. Can you share your codes or ideas with us so that we can find out why you need to change `MODEL`'s build function? If it is necessary, we may consider changing the logic here.

**Reply by Fei-Wang** [reply]
I want to use in my nlp project, and I use the popular nlp library called transformers, here's my code in my build_func,
```python3
    pretrained = args.pop('pretrained')
    if pretrained:
        obj = obj_cls.from_pretrained(pretrained, **args)
    else:
        obj = obj_cls(**args)  # type: ignore

    return obj
```
so, I can init the model use from_pretrained method directly, of course, my model is inherited from transformers.PreTrainedModel and BaseModel from mmengine.
but I meet some problems, and I'm thinking about solving it.
1. use my own ```build_func```
2. ```self._init_model_weights()``` repeated initialization in runner.train()
3. save and load model use save_pretrained and from_pretrained
4. add save and load other staff such as tokenizer...

My current idea is to rewrite the runner, but in this way, I will not enjoy the advantages and convenience brought by mmengine. Is there a better way to solve it?

**Reply by C1rN09** [reply]
For Q1, you can actually register a function to your own `Registry`. For example

```python
# suppose MODELS is your Registry rather than MMEngine's root Registry
@MODELS.register_module(name='MyNLPModel')
def build_my_nlp_model(*args, **kwargs):
    # args and kwargs are your own parameters in config files, you can make your own specification
    # However, `type` filed has been popped in MODELS.build(), so you will have to use some other names
    pass

# In your config file, you should write
model = dict(type='MyNLPModel', ...)
```

Actually, this is a "build function in build function". Feel free to give us more feedback if you are confused, or you have any suggestions!

**Reply by C1rN09** [reply]
For Q2, you can manually remove the `init_weights()` method inherited from `BaseModel`

**Reply by C1rN09** [reply]
For Q3/Q4, we are considering making save/load checkpoints as `BaseModel's` method so that you can modify it on your need. This is under investigation and may not come out soon.
